Classes/ProcessManager/AbstractProcessManager.php

```diff
@@ -0,0 +1,87 @@
+use Neos\Cqrs\EventStore\RawEvent;
+use Neos\Cqrs\ProcessManager\State\State;
+use Neos\Cqrs\ProcessManager\State\StateRepository;
+use TYPO3\Flow\Annotations as Flow;
+
+/**
+ * Base implementation for a process manager
+ */
+abstract class AbstractProcessManager implements EventListenerInterface, EventListenerInvocationInterface
+{
+    /**
+     * @Flow\Inject
+     * @var StateRepository
+     */
+    protected $stateRepository;
+
+    /**
+     * @var State
+     */
+    protected $state;
+
+    /**
+     * Return the process configuration for the concrete process manager
+     *
+     * This method must return an array which contains (short) event names and a correlating closure which
+     * returns a process state identifier. This process state identifier will be used throughout all event listener
+     * methods in the process manager to always retrieve the correct state.
+     *
+     * The "short event name" is the same used in the "when<ShortName>()" method name of the event listener method.
+     *
+     * The following example uses the organization's identifier as the process state identifier:
+     *
+     *     return [
+     *         'OrganizationHasBeenCreated' => function(OrganizationHasBeenCreated $event) {
+     *             return $event->getIdentifier();
+     *         },
+     *         'MemberHasBeenAddedToOrganization' => function(MemberHasBeenAddedToOrganization $event) {
+     *             return $event->getOrganizationIdentifier();
+     *         },
+     *         'NewDeploymentKeyHasBeenGenerated' => function(NewDeploymentKeyHasBeenGenerated $event) {
+     *             return $event->getOrganizationIdentifier();
+     *         }
+     *
+     * @return array
+     */
+    abstract protected function getProcessConfiguration(): array;
+
+    /**
+     * Prepare the process state and then call the intended event listener method
+     *
+     * @param string $eventListenerMethodName
+     * @param EventInterface $event
+     * @param RawEvent $rawEvent
+     * @throws ProcessManagerException
+     */
+    public function invokeEventListenerMethod(string $eventListenerMethodName, EventInterface $event, RawEvent $rawEvent)
+    {
+        $configuration = $this->getProcessConfiguration();
+        $eventName = substr($eventListenerMethodName, 4);
+        if (!isset($configuration[$eventName])) {
+            throw new ProcessManagerException(sprintf('The event listener method "%s" does not have a matching process configuration with key "%s". Check the array returned by getProcessConfiguration() in class %s.', $eventListenerMethodName, $eventName, get_class($this)), 1479290834150);
+        }
+
+        $stateIdentifier = $configuration[$eventName]($event);
+        $this->state = $this->stateRepository->get($stateIdentifier, get_class($this));
+        if ($this->state === null) {
+            $this->state = new State($stateIdentifier, get_class($this));
+        }
```

Comment:
As far as I can see, that wouldn't help in my case because I need to prepare the process state individually for each `when*()` call, not only once when the process manager is instantiated. And furthermore, my process manager currently is a singleton (because I can! ;-) ).